configmap/dynamic.go

```diff
@@ -0,0 +1,250 @@
+	Close()
+}
+
+type dynamicConfigInternal struct {
+	// Parameters
+	configMapName string
+	options       Options
+	configMapper  ConfigMapper
+	// Internal
+	client        *kubernetes.Clientset
+	lock          sync.Mutex
+	refreshCancel chan struct{}
+	// Recently processed data
+	bootstrap *ConfigState
+	current   *ConfigState
+}
+
+func NewDynamicConfig(config *rest.Config, configMapName string, configMapper ConfigMapper, options Options) (DynamicConfig, error) {
+	internal := dynamicConfigInternal{
+		client:        kubernetes.NewForConfigOrDie(config),
+		configMapName: configMapName,
+		options:       options,
+		configMapper:  configMapper,
+	}
+	data, err := internal.loadAndTryFixConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	if err = internal.process(data); err != nil {
+		return nil, err
+	}
+	internal.bootstrap = internal.current
+
+	if !options.DisableDynamicUpdates {
+		internal.startRefreshProcess(kubernetes.NewForConfigOrDie(config))
+	}
+
+	return &internal, nil
+}
+
+func NewDynamicConfigFromMap(data map[string]string, configMapper ConfigMapper, options Options) (DynamicConfig, error) {
+	internal := dynamicConfigInternal{
+		options:      options,
+		configMapper: configMapper,
+	}
+	if err := internal.process(data); err != nil {
+		return nil, err
+	}
+	internal.bootstrap = internal.current
+	return &internal, nil
+}
+
+func (d *dynamicConfigInternal) GetBootstrap() interface{} {
+	return d.bootstrap.Mapped
+}
+
+func (d *dynamicConfigInternal) GetRawBootstrap() map[string]string {
+	return d.bootstrap.Raw
+}
+
+func (d *dynamicConfigInternal) Get() interface{} {
+	return d.getCurrent().Mapped
+}
+
+func (d *dynamicConfigInternal) GetRaw() map[string]string {
+	return d.getCurrent().Raw
+}
+
+func (d *dynamicConfigInternal) Close() {
+	if d.refreshCancel != nil {
+		d.refreshCancel <- struct{}{}
+	}
+}
+
+func (d *dynamicConfigInternal) getCurrent() *ConfigState {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	return d.current
+}
+
+func (d *dynamicConfigInternal) loadAndTryFixConfig() (map[string]string, error) {
+	configMapsClient := d.client.CoreV1().ConfigMaps("default")
+	controllerConfig, err := configMapsClient.Get(context.TODO(), d.configMapName, metaV1.GetOptions{})
+	if err != nil {
+		if d.options.Template == nil {
+			return nil, err
+		}
+		if serr, ok := err.(*errors.StatusError); !ok || serr.ErrStatus.Reason != metaV1.StatusReasonNotFound {
+			return nil, err
+		}
+		configMap := corev1.ConfigMap{
+			ObjectMeta: metaV1.ObjectMeta{
+				Namespace: "default",
+				Name:      d.configMapName,
+			},
+			Data: d.options.Template,
+		}
+		createdConfigMap, createErr := configMapsClient.Create(context.TODO(), &configMap, metaV1.CreateOptions{})
+		if createErr != nil {
+			return nil, createErr
+		}
+		return createdConfigMap.Data, err
+	}
+	data := d.updateConfigMap(configMapsClient, controllerConfig)
+	return data, nil
+}
+
+func (d *dynamicConfigInternal) updateConfigMap(configMapsClient v1.ConfigMapInterface,
```

Comment:
This is really `updateConfigMapWithDefaultsFromTemplate` right?
It looks like it goes through the template and sets things that are missing